internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go

```diff
@@ -214,6 +255,119 @@ func (f *Fetcher) FetchDestinationsData(
+	}
+
+	// For now, we need to look up all service endpoints within a partition.
+	rsp, err := f.Client.List(ctx, &pbresource.ListRequest{
+		Type: catalog.ServiceEndpointsType,
+		Tenancy: &pbresource.Tenancy{
+			Namespace: storage.Wildcard,
+			Partition: proxyID.Tenancy.Partition,
+			PeerName:  proxyID.Tenancy.PeerName,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range rsp.Resources {
+		// If it's already in destinations, ignore it.
+		if _, ok := destinations[resource.NewReferenceKey(r.Id)]; ok {
+			continue
+		}
+
+		var endpoints pbcatalog.ServiceEndpoints
+		err = r.Data.UnmarshalTo(&endpoints)
+		if err != nil {
+			return nil, err
+		}
+
+		// If this proxy is a part of this service, ignore it.
+		if isPartOfService(resource.ReplaceType(catalog.WorkloadType, proxyID), &endpoints) {
+			continue
+		}
+
+		// Collect all identities.
+		var identities []*pbresource.Reference
+		for _, ep := range endpoints.Endpoints {
+			identities = append(identities, &pbresource.Reference{
+				Name:    ep.Identity,
+				Tenancy: r.Id.Tenancy,
+			})
+		}
+
+		// Fetch the service.
+		// todo (ishustava): this should eventually grab virtual IPs resource.
+		s, err := f.FetchService(ctx, resource.ReplaceType(catalog.ServiceType, r.Id))
+		if err != nil {
+			return nil, err
+		}
+		if s == nil {
+			// If service no longer exists, skip.
+			continue
+		}
+
+		d := &intermediateTypes.Destination{
+			ServiceEndpoints: &intermediateTypes.ServiceEndpoints{
+				Resource:  r,
+				Endpoints: &endpoints,
+			},
+			VirtualIPs: s.Service.VirtualIps,
+			Identities: identities,
+		}
+		addToDestinations = append(addToDestinations, d)
+	}
+	return addToDestinations, err
+}
+
+// FetchAndMergeProxyConfigurations fetches proxy configurations for the proxy state template provided by id
+// and merges them into one object.
+func (f *Fetcher) FetchAndMergeProxyConfigurations(ctx context.Context, id *pbresource.ID) (*pbmesh.ProxyConfiguration, error) {
+	proxyCfgRefs := f.ProxyCfgCache.ProxyConfigurationsByProxyID(id)
+
+	result := &pbmesh.ProxyConfiguration{
+		DynamicConfig: &pbmesh.DynamicConfig{},
+	}
+	for _, ref := range proxyCfgRefs {
+		proxyCfgID := &pbresource.ID{
+			Name:    ref.GetName(),
+			Type:    ref.GetType(),
+			Tenancy: ref.GetTenancy(),
+		}
+		rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{
+			Id: proxyCfgID,
+		})
+		switch {
+		case status.Code(err) == codes.NotFound:
+			f.ProxyCfgCache.UntrackProxyConfiguration(proxyCfgID)
+			return nil, nil
+		case err != nil:
+			return nil, err
+		}
+
+		var proxyCfg pbmesh.ProxyConfiguration
+		err = rsp.Resource.Data.UnmarshalTo(&proxyCfg)
+		if err != nil {
+			return nil, err
+		}
+
+		// Note that we only care about dynamic config as bootstrap config
+		// will not be updated dynamically by this controller.
+		// todo (ishustava): do sorting etc.
+		proto.Merge(result.DynamicConfig, proxyCfg.DynamicConfig)
```

Comment:
It looks like the bimapper iterates over a map internally, so I bet `ProxyConfigurationsByProxyID`'s return value and this code is indeterministic.

Reply:
If values don't conflict, we don't care. If values are conflicting we care, but I left a TODO that this will come later (what I meant by `sorting etc` here 😄 )